examples/autoscaling/src/main/java/autoscaling/LoadSimulationPipeline.java

```diff
@@ -0,0 +1,184 @@
+ *                    in the sense that at least maxLoad amount of subtasks are needed to sustain the load.
+ *                    For example, a max load of 1 represents 100% load on a single subtask, 50% load on two subtasks.
+ *                    Similarly, a max load of 2 represents 100% load on two tasks, 50% load on 4 subtasks.
+ *
+ *                    Multiple tasks and branches can be defined to test Flink Autoscaling. The format is as follows:
+ *                       maxLoadTask1Branch1[;maxLoadTask2Branch1...[\n maxLoadTask1Branch2[;maxLoadTask2Branch2...]...]
+ *
+ *                    A concrete example: "1;2;4\n4;2;1"
+ *                      Two branches are created with three tasks each. On the first branch, the tasks have
+ *                      a load of 1, 2, and 3 respectively. On the second branch, the tasks have the load reversed.
+ *                      This means, that at peak Flink Autoscaling at target utilization of 0.5, the parallelisms of
+ *                      the tasks will be 2, 4, 8 for branch one and vise-versa for branch two.
+ * </pre>
+ */
+public class LoadSimulationPipeline {
+
+    private static final Logger LOG = LoggerFactory.getLogger(LoadSimulationPipeline.class);
+
+    public static void main(String[] args) throws Exception {
+        var env = StreamExecutionEnvironment.getExecutionEnvironment();
+        env.disableOperatorChaining();
+
+        var arguments = ParameterTool.fromArgs(args);
+        String maxLoadPerTask =
+                arguments.get("maxLoadPerTask", "1;2;4;8;16;\n16;8;4;2;1\n8;4;16;1;2");
+        long repeatsAfterMs =
+                Duration.ofMinutes(arguments.getLong("repeatsAfterMinutes", 60)).toMillis();
+        int samplingIntervalMs = arguments.getInt("samplingIntervalMs", 1_000);
+
+        for (String branch : maxLoadPerTask.split("\n")) {
+            String[] taskLoads = branch.split(";");
+
+            DataStream<Long> stream =
+                    env.addSource(new ImpulseSource(samplingIntervalMs)).name("ImpulseSource");
+
+            for (String load : taskLoads) {
+                double maxLoad = Double.parseDouble(load);
+                stream =
+                        stream.shuffle()
+                                .flatMap(
+                                        new LoadSimulationFn(
+                                                maxLoad, repeatsAfterMs, samplingIntervalMs))
+                                .name("MaxLoad: " + maxLoad)
+                                .broadcast();
+            }
+
+            stream.addSink(new DiscardingSink<>());
+        }
+
+        env.execute(
+                "Load Simulation (repeats after "
+                        + Duration.of(repeatsAfterMs, ChronoUnit.MILLIS)
+                        + ")");
+    }
+
+    private static class ImpulseSource implements SourceFunction<Long> {
+        private final int maxSleepTimeMs;
+        boolean canceled;
```

Comment:
has to be `volatile`

Reply:
I'm not sure about that. This will still stop, there can be a small delay though.